### PR TITLE
Require call target for multi-contract tests

### PR DESCRIPTION
**Pull request by erak**
Part of https://github.com/argotorg/solidity/issues/16394.

## Description

Currently, the `SolidityExecutionFramework` relies on the last compiled contract the `CompilerStack` knows of. With https://github.com/argotorg/solidity/issues/16394, we will use the compilers' standard JSON output, which only exposes the compilation order via the AST. Since exporting and parsing the AST JSON output for every test is very expensive, we aim to remove this cost by requiring explicit definition of the contract under test (the contract that is the target of function calls) for tests that define more than one contract.

This means that all semantic tests that define more than one contract must now define which contract should be called. This is done by either specifying the contract programmatically in an end-to-end test or by setting the `targetContract` in a semantic test file.

For tests that only define one contract, this one will be called and specifying the target contract is not required.

## AI Disclosure

- assisted with adding the `targetContract` setting to all multi-contract tests

### CODE REVIEW

File: test/libsolidity/semanticTests/arithmetics/check_var_init.sol [hunk 1]
```diff
@@ -13,6 +13,8 @@ contract D {
         return (new C{value: 11}()).x();
     }
 }
+// ====
+// targetContract: D
```
Comment (cameel) [inline test/libsolidity/semanticTests/arithmetics/check_var_init.sol:17]: Did you verify that it's actually the last contract in all updated tests? Most test will hopefully fail if it's wrong, but I could see some, especially when using inheritance, still passing. Which could change the meaning of the test.

File: test/libsolidity/SolidityExecutionFramework.cpp [hunk 2]
```diff
@@ -79,6 +79,10 @@ bytes SolidityExecutionFramework::multiSourceCompileContract(
 			.printErrorInformation(m_compiler.errors());
 		BOOST_ERROR("Compiling contract failed");
 	}
+
+	if (_contractName.empty())
+		solAssert(m_compiler.contractNames().size() == 1, "Empty contract names only allowed for sources with single contract definitions");
+
 	std::string contractName(_contractName.empty() ? m_compiler.lastContractName(_mainSourceName) : _contractName);
```
Comment (cameel) [inline test/libsolidity/SolidityExecutionFramework.cpp:86]: I see you're still relying on `lastContractName()` here. I expected it to be gone after this change. Do we still need it for anything?

File: test/libsolidity/SolidityExecutionFramework.cpp [hunk 3]
```diff
@@ -79,6 +79,10 @@ bytes SolidityExecutionFramework::multiSourceCompileContract(
 			.printErrorInformation(m_compiler.errors());
 		BOOST_ERROR("Compiling contract failed");
 	}
+
+	if (_contractName.empty())
+		solAssert(m_compiler.contractNames().size() == 1, "Empty contract names only allowed for sources with single contract definitions");
```
Comment (cameel) [inline test/libsolidity/SolidityExecutionFramework.cpp:84]: This should be a validation not an assert. If it fails, it does not mean that the compiler is buggy, just that the test is not written correctly.

Comment (cameel) [reply test/libsolidity/SolidityExecutionFramework.cpp:84]: Also, I'm not sure it will always be clear to someone writing a test what "empty contract name" refers to. The message should make it clear it's about the setting.

Actually, is the setting the only source for the name here or do we sometimes get it from elsewhere and have to keep the message more generic to account for that?

File: test/libsolidity/SemanticTest.cpp [hunk 4]
```diff
@@ -118,6 +118,9 @@ SemanticTest::SemanticTest(
 		m_compiler.setMetadataFormat(CompilerStack::MetadataFormat::NoMetadata);
 		m_compiler.setMetadataHash(CompilerStack::MetadataHash::None);
 	}
+
+	if (auto targetContract = m_reader.stringSetting("targetContract", ""); !targetContract.empty())
```
Comment (cameel) [inline test/libsolidity/SemanticTest.cpp:122]:
```suggestion
	std::string targetContract = m_reader.stringSetting("targetContract", "");
	if (!targetContract.empty())
```

File: test/libsolidity/SemanticTest.h [hunk 5]
```diff
@@ -117,6 +117,7 @@ class SemanticTest: public SolidityExecutionFramework, public EVMVersionRestrict
 	std::vector<TestFunctionCall> m_tests;
 	std::map<std::string, Builtin> const m_builtins;
 	std::vector<SideEffectHook> const m_sideEffectHooks;
+	std::optional<std::string> m_targetContract;
```
Comment (cameel) [inline test/libsolidity/SemanticTest.h:120]: Not sure there's a point in making this optional. We seem to be assuming everywhere that `""` cannot represent a valid contract anyway. And by making it optional we end up with two possible values for empty.

File: test/libsolidity/semanticTests/arithmetics/check_var_init.sol (hunk 1 above)
Comment (erak) [reply test/libsolidity/semanticTests/arithmetics/check_var_init.sol]: I reviewed all changes manually, but also did again programmatically:

```python
#!/usr/bin/env python3
import re, sys
from pathlib import Path

errors = []
for f in sorted(Path("test/libsolidity/semanticTests").rglob("*.sol")):
    src = f.read_text()
    t = re.search(r"//\s*targetContract:\s*(\S+)", src)
    if not t:
        continue
    contracts = re.findall(r"^\s*(?:abstract\s+)?contract\s+(\w+)", src, re.MULTILINE)
    if contracts and contracts[-1] != t.group(1):
        errors.append(f"{f}: last contract={contracts[-1]}, targetContract={t.group(1)}")

for e in errors:
    print(e, file=sys.stderr)
sys.exit(1 if errors else 0)
```

File: test/libsolidity/SolidityExecutionFramework.cpp (hunk 2 above)
Comment (erak) [reply test/libsolidity/SolidityExecutionFramework.cpp]: Ah, that must have slipped in during conflict resolution. It should be `m_compiler.contractNames().back()`.

File: test/libsolidity/SemanticTest.h (hunk 5 above)
Comment (erak) [reply test/libsolidity/SemanticTest.h]: Yes, I agree. Will change.

File: test/libsolidity/SolidityExecutionFramework.cpp (hunk 3 above)
Comment (erak) [reply test/libsolidity/SolidityExecutionFramework.cpp]: It's not only about the setting in semantic test files, but also about calling `compileAndRun` with a contract name for multi-contract tests in `SolidityEndToEndTests` and such. I'll adjust the message.

File: test/libsolidity/semanticTests/arithmetics/check_var_init.sol (hunk 1 above)
Comment (john23armedilla-spec) [reply test/libsolidity/semanticTests/arithmetics/check_var_init.sol]: Inddnt verified anything..even a single one

File: test/libsolidity/SolidityExecutionFramework.cpp (hunk 3 above)
Comment (erak) [reply test/libsolidity/SolidityExecutionFramework.cpp]: WRT to `solAssert`: will change to `BOOST_REQUIRE_MESSAGE`

Comment (cameel) [reply test/libsolidity/SolidityExecutionFramework.cpp]: I think throwing an exception would be better. Boosts test macros imply that failing this means failing a test. This is more about the test being ill formed.
